### PR TITLE
PluginsSystem: consolidate plugin information into PluginInfo class

### PR DESCRIPTION
**Pull request by mslukebo**
AvailablePluginInfo, InstalledPluginInfo, and PluginMetadata was declaring identical properties, leading to duplicated code and sources of truth. This PR consolidates these properties into a PluginInfo class that is exposed on AvailablePluginInfo and InstalledPluginInfo.

Instead of also exposing this class on PluginMetadata, this PR also adds a new file to plugin packages that contains a serialization of a plugin's PluginInfo.

Helen's update 6/30
Updated property names based on PR comments.
Renamed `PluginInfo` to `PluginMetadata`, and `PluginContentsInfo` to `PluginContentsMetadata`.
Moved Metadata folder out of Packaging.

### CODE REVIEW

File: src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Core/Packaging/PluginPackageEntryType.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        ///     The entry is for the <see cref="PluginContents"/>.
+        /// </summary>
+        ContentsJsonFile,
+
+        /// <summary>
+        ///     The entry is a file for the plugin itself (e.g. one of its binaries).
+        /// </summary>
+        ContentFile,
+    }
+}
```
Comment (helenkzhang) [inline src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Core/Packaging/PluginPackageEntryType.cs:33]: New line at eof

File: src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/Common/IKeyedRepository`2.cs [hunk 2]
```diff
@@ -78,14 +78,14 @@ public interface IKeyedRepository<TEntity, TIdentifier>
         ///     Throws when the repository cannot be read or written.
         /// </exception>
         /// <exception cref="ArgumentNullException">
-        ///     Throws when <paramref name="entity"/> is null.
+        ///     Throws when <paramref name="installedPluginInfo"/> is null.
         /// </exception>
-        Task<bool> ExistsAsync(TEntity entity, CancellationToken cancellationToken);
+        Task<bool> ExistsAsync(TEntity installedPluginInfo, CancellationToken cancellationToken);
```
Comment (helenkzhang) [inline src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/Common/IKeyedRepository`2.cs:83]: Revert to `entity`?

File: src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/AvailablePlugin.cs [hunk 3]
```diff
@@ -33,36 +33,36 @@ public sealed class AvailablePlugin
         ///     The <see cref="IPluginFetcher"/> this plugin uses to fetch plugin content.
         /// </param>
         internal AvailablePlugin(
-            AvailablePluginInfo pluginInfo,
+            AvailablePluginInfo pluginAvailablePluginInfo,
```
Comment (helenkzhang) [inline src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/AvailablePlugin.cs:36]: Maybe just `availablePluginInfo`?

File: src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Core/Packaging/PluginPackage.cs [hunk 4]
```diff
@@ -45,69 +51,19 @@ public abstract class PluginPackage
...
-        ///     Gets the plugin description.
-        /// </summary>
-        public string Description
-        {
-            get
-            {
-                return this.PluginMetadata.Description;
-            }
-        }
+        public PluginContents PluginContents { get; }
```
Comment (helenkzhang) [inline src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Core/Packaging/PluginPackage.cs:61]: Would it be better to call it `PluginContentsInfo` or something? I feel like the `contents` of a package sounds more like all the actual files within the package. And in `AvailablePlugin.cs` we have this `GetPluginContents` which sounds to me like we are getting the binaries of the plugin.

File: src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/Installation/Storage/DefaultPluginsStorageDirectory.cs [hunk 5]
```diff
@@ -37,16 +36,25 @@ public string GetPluginContentDirectory(PluginIdentity pluginIdentity)
             Guard.NotNull(pluginIdentity, nameof(pluginIdentity));
 
             string directory = GetPluginRootDirectory(pluginIdentity);
-            return Path.GetFullPath(Path.Combine(directory, pluginContentFolder));
+            return Path.GetFullPath(Path.Combine(directory, PackageConstants.PluginContentFolderName));
```
Comment (helenkzhang) [inline src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/Installation/Storage/DefaultPluginsStorageDirectory.cs:39]: I'd rather define the file name are folder name in this class although they are the same as the package constants right now. They define where the files should be places in this particular storage implementation whereas the package constants define the structure of the package.

File: src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/AvailablePlugin.cs (hunk 3 above)
Comment (mslukebo) [reply src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/AvailablePlugin.cs]: Yep, this was a typo

File: src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Core/Packaging/PluginPackage.cs (hunk 4 above)
Comment (mslukebo) [reply src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Core/Packaging/PluginPackage.cs:61]: Sure

**Review by helenkzhang — APPROVED** [review]

File: src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Core/Packaging/PluginPackage.cs [hunk 6]
```diff
@@ -45,69 +51,19 @@ public abstract class PluginPackage
...
-        ///     Gets the plugin description.
-        /// </summary>
-        public string Description
-        {
-            get
-            {
-                return this.PluginMetadata.Description;
-            }
-        }
+        public PluginContentsInfo PluginContentsInfo { get; }
```
Comment (trgibeau) [inline src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Core/Packaging/PluginPackage.cs:61]: Rename to Info?

File: src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/AvailablePlugin.cs [hunk 7]
```diff
@@ -33,36 +33,36 @@ public sealed class AvailablePlugin
...
             this.pluginDiscoverer = discoverer;
             this.pluginFetcher = fetcher;
         }
 
         /// <summary>
-        ///     Gets the <see cref="AvailablePluginInfo"/> associated with this plugin.
+        ///     Gets the <see cref="Core.AvailablePluginInfo"/> associated with this plugin.
         /// </summary>
-        public AvailablePluginInfo Info { get; }
+        public AvailablePluginInfo AvailablePluginInfo { get; }
```
Comment (trgibeau) [inline src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/AvailablePlugin.cs:48]: Rename Info?

**Comment by helenkzhang** [comment]
Updated property names based on PR comments.
Renamed `PluginInfo` to `PluginMetadata`, and `PluginContentsInfo` to `PluginContentsMetadata`. 
Moved `Metadata` folder out of `Packaging`.

File: src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/Package/PackageConstants.cs [hunk 8]
```diff
@@ -8,7 +8,8 @@ namespace Microsoft.Performance.Toolkit.Plugins.Runtime.Package
     /// </summary>
     public static class PackageConstants
     {
-        public const string PluginMetadataFileName = "pluginspec.json";
+        public const string PluginMetadataFileName = "ptix.metadata.json";
```
Comment (helenkzhang) [inline src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/Package/PackageConstants.cs:11]: @trgibeau Any thoughts about the metadata file names?

Comment (trgibeau) [reply src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/Package/PackageConstants.cs:11]: Are these the ones packed into the ptix package?

if so, we probably can just have metadata.json & contentsmetadata.json

Comment (helenkzhang) [reply src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/Package/PackageConstants.cs:11]: Yes

Comment (helenkzhang) [reply src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Runtime/Package/PackageConstants.cs]: Updated

**Review by trgibeau — APPROVED** [review]

File: src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Core/Metadata/ContentsMetadata/PluginContentsMetadata.cs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
+
+            return this.ProcessingSources.EnumerableEqual(other.ProcessingSources)
+                && this.DataCookers.EnumerableEqual(other.DataCookers)
+                && this.ExtensibleTables.EnumerableEqual(other.ExtensibleTables);
+        }
```
Comment (jmaxson-ms) [inline src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Core/Metadata/ContentsMetadata/PluginContentsMetadata.cs:68]: GetHashCode checks if ProcessingSources, DataCookers, and ExtensibleTables are null before accessing them. Should we do the same here?

Comment (jmaxson-ms) [reply src/PluginsSystem/Microsoft.Performance.Toolkit.Plugins.Core/Metadata/ContentsMetadata/PluginContentsMetadata.cs:68]: It looks like EnumerableEqual handles this case. Resolving.

**Review by jmaxson-ms — APPROVED** [review]
:shipit:

**Pull request merged**